Refactor

JamesTFarrington opened the pull request:
simplifying some things

JamesTFarrington commented:
I have this build passing with an updated pypy. The worker functionality is working great, except for each worker returning an import error.

jMyles commented:
@JamesTFarrington, what more can you tell us about this work?  Can you describe the 2-3 most important changes you've made to the worker logic?

JamesTFarrington commented:
1.Starting Hendrix... and Stopping Hendrix.... messages now print when HendrixDeploy is started or stopped, rather when start or stop are parsed as input.

2. Removed daemonize. Our previous method was not a proper unix daemon. I tried using python-daemon which would cause ImportError's and/or cause twisted to have i/o errors. I also looked at how twisted does daemons, which doesn't seem to work within python, you have to call it from bash.
I think the solution here is going to be to encourage the users to install hendrix as a service.

3.Refactor workers.
I removed parts of launchWorkers that were not specific to workers functionality, and wrote the openPidList method. I removed launchWorkers from HendrixDeploy and made it its own function. I subclassed HendrixDeploy, and seperated anything that only pertains to workers (I seem to have missed getSpawnArgs).
We check for workers in ux, before any instance of HendrixDeploy is initialized
If workers is not 0 deployWorkers is called. This was launching hx via reactor.spawnProcess. (after a few attempts to find an alternative to reactor.spawnProcess) I wrote a new script 'hxw' which called ux.launch directly. After some trial and error I wrote a new function to parse arguments from hxw

jMyles commented:
@JamesTFarrington, amazing!

So, off the bat, I think we have to conclude that merging this puts us at 3.x, right?  It clearly introduces a backward-incompatibility against a feature that is in use (albeit, probably needs to not be in use) in at least some of the current 2.x deployments.

I'm also wondering: what, if anything, can we do to force this to pass on Travis?  Can we detect the pypy version and refrain from running the test(s) that import cryptography?

JamesTFarrington commented:
I don't think we should rush to move to 3.x so fast. We could probably have the parser print something like "-d is being deprecated soon, try using & or installing hendrix as a service." Also, this is still having an issue with workers having an import error.

As far as getting this to pass on travis, travis ci suggests switching to their docker based infrastructure. I'm not sure if that would do anything. I don't think we can get it to pass by refraining from tests. Travis ci has an error when it tries to install cryptography when it is setting up. So, it's failing before it starts to run tests.

jMyles commented:
Looking better all the time.

So, as I said in the slack, I don't think there's any reason to be slow about getting to 3.x.  On the contrary: if there are backward-incompatible changes that are awesome and can land soon, let's get 3.x out the door right away, before even more people have this thing in production.

As for pypy: I see that you have removed it for now to get it to pass.  Reasonable enough.  However, the "dockerized" solution seems like a great idea to me - that will empower us to run the performance tests as well and report the results in a fair manner.

What does this entail?  @dmpayton, have you any experience with this?

JamesTFarrington commented:
It looks like I miss spoke. It's "container-based" not dockerized.
